src/Lykke.AlgoStore.Services/Validation/CSharpCodeBuildSession.cs

```diff
@@ -0,0 +1,309 @@
+
+        public async Task<AlgoMetaDataInformation> ExtractMetadata()
+        {
+            var root = (CompilationUnitSyntax)await _syntaxTree.GetRootAsync();
+            var namespaceDeclaration = (NamespaceDeclarationSyntax)root.Members[0];
+            var classDeclaration = namespaceDeclaration.Members
+                .OfType<ClassDeclarationSyntax>()
+                .First();
+
+            var newAssembly = GenerateAssembly();
+            var newType = newAssembly.GetType($"{namespaceDeclaration.Name}.{classDeclaration.Identifier.Text}");
+            var metadata = ExtractMetadata(newType);
+
+            return metadata;
+        }
+
+        private AlgoMetaDataInformation ExtractMetadata(Type algoType)
+        {
+            var metadata = new AlgoMetaDataInformation
+            {
+                Functions = new List<AlgoMetaDataFunction>(),
+                Parameters = new List<AlgoMetaDataParameter>()
+            };
+
+            var algoProperties = algoType
+                .GetProperties(BindingFlags.Instance | BindingFlags.Public);
+
+            foreach (var algoProperty in algoProperties)
+            {
+                var parameter = ToAlgoMetaDataParameter(algoProperty);
+
+                if (algoProperty.PropertyType.IsEnum)
+                    parameter.PredefinedValues = ToEnumValues(algoProperty);
+
+                metadata.Parameters.Add(parameter);
+            }
+
+            var functionProperties = algoType
+                .GetRuntimeFields()
+                .Where(x => x.FieldType.BaseType == typeof(AbstractFunction));
+
+            foreach (var functionProperty in functionProperties)
+            {
+                var function = ToAlgoMetadataFunction(functionProperty);
+
+                var functionParamBasePropertyField = functionProperty
+                    .FieldType
+                    .GetFields()
+                    .FirstOrDefault(x => x.FieldType.BaseType == typeof(FunctionParamsBase));
+
+                //There is a public property which base type is of FunctionParamBase type
+                if (functionParamBasePropertyField != null)
+                {
+                    function.FunctionParameterType = functionParamBasePropertyField.FieldType.FullName;
+                    var functionParameters = functionParamBasePropertyField.FieldType.GetProperties();
+
+                    foreach (var functionParameter in functionParameters)
+                    {
+                        var parameter = ToAlgoMetaDataParameter(functionParameter);
+
+                        if (functionParameter.PropertyType.IsEnum)
+                            parameter.PredefinedValues = ToEnumValues(functionParameter);
+
+                        function.Parameters.Add(parameter);
+
+                    }
+                }
+                //There is a NO public property which base type is of FunctionParamBase type,
+                //so we should get base algo parameters
+                else
```

Comment:
Do we need to look at both fields and properties? I think this should be consistent across all searches. For example, if we search only for properties, algo params should also be looked for only in properties.

Also, most of the code in this else block is duplicated with the code above